Add system module with push_env and push_argv

kgriffs opened the pull request:
This patch factors out push_env and push_argv from Olivaw. These context managers are useful in testing CLI apps and also for passing options to scripts when you can't directly control the way they are subprocessed (e.g., Ansible dynamic inventory scripts).

Commit:
Add system module with push_env and push_argv

This patch factors out push_env and push_argv from Olivaw. These
context managers are useful in testing CLI apps and also for
passing options to scripts when you can't directly control the way
they are subprocessed (e.g., Ansible dynamic inventory scripts).

lvh commented on rmsutil/system.py:
```diff
@@ -0,0 +1,31 @@
+@contextlib.contextmanager
+def push_env(**kwargs):
+    try:
+        old_env = os.environ.copy()
```
Warning:  `os.environ` is an instance of a subclass of dict, and subclassing is of course anathema, no good, super bad, etc.

`os.environ` is an `instance` (yep, really), `os.environ.copy()` is a `dict`. If you want to build your own environ, you might want to consider `os._Environ` because that's what `os.environ` is actually an instance of.

lvh replied:
I guess this is incidentally OK here, because you're never reassigning `os.environ`, but that should probably come with a warning in the code.

lvh commented on rmsutil/system.py:
```diff
@@ -0,0 +1,31 @@
+
+@contextlib.contextmanager
+def push_argv(*args):
+    old_argv = copy.copy(sys.argv)
```
Ignorable nitpick: since you're slicing sys.argv, you're getting a new list, so I don't think you need a copy.

kgriffs replied:
good point

lvh commented:
LGTM! Clean, great tests. Might wanna add a comment about that os.environ landmine.

Commit:
Add test for push_env side effects

kgriffs commented:
Added comments and docstrings, removed copy.copy in push_argv, and added one more check in test_push_env to assert that a value can be overridden and restored.

I've got a few more tests I'll push in just a minute, then I think this is good to go for final review.

Commit:
Require string values in push_argv, push_env

lvh commented on rmsutil/system.py:
```diff
@@ -0,0 +1,66 @@
+        raise TypeError('All args must be strings')
+
+    old_argv, sys.argv = sys.argv, sys.argv[:1]
+    sys.argv.extend(args)
```
_blink_

I'm not sure how this works. It appears to throw away the process name (`sys.argv[0]`)? Why does it do that?

Regardless, perhaps consider:

```
old_argv, sys.argv = sys.argv, sys.argv[:1] + args
```

... since IMHO that makes it more clear what's going on than `extend` does.

lvh commented:
LGTM modulo that thing I don't understand :D

Commit:
Rename push_env and push_argv for clarity
